### PR TITLE
MDEV-35747: Wrong result from prepared TVC with parameter markers

### PR DESCRIPTION
**Pull request by DaveGosselin-MariaDB**
The setup of column type information in table_value_constr::prepare() was wrapped in an "if (!holders)" guard so that it runs only once per prepared statement.  However, the guard was too wide because it bound the allocation of item holders (which should happen only once) to the collection of type information (which should happen on each execution).

This leaves the TVC stuck with whatever placeholder type the parameter had at PREPARE time which likely won't match the type of the next substitution (because a type holder has no actual type at PREPARE time).  Its type only becomes known when a value is bound at EXECUTE time.  So both the TVC types and the corresponding Item_type_holder instance in the SELECT item list must be computed again on every EXECUTE.

This patch does just that, and separates the work done once per prepared statement from the work done on every execution as implied above.

### CODE REVIEW

File: sql/sql_class.h [hunk 1]
```diff
@@ -8249,6 +8249,13 @@ class Type_holder: public Sql_alloc,
   bool aggregate_attributes(THD *thd)
   {
     static LEX_CSTRING union_name= { STRING_WITH_LEN("UNION") };
+    /*
+      m_maybe_null is OR-accumulated and never reset here, so it is
+      sticky across calls.  In a prepared statement context this means
+      that once any EXECUTE has observed a NULL for a given column, that
+      column stays nullable for the rest of the statement's lifetime
+      even if later EXECUTEs only bind non-NULL values.
+    */
```
[the remark this answers is not in the document]
Comment (DaveGosselin-MariaDB) [reply sql/sql_class.h:8270]: Are you sure?  The aggregate_attributes_* helpers all assign max_length, decimals, unsigned_flag, collation directly (not |=); aggregate_numeric_attributes_real even explicitly zeroes them before the loop.  Only m_maybe_null is OR-accumulated (which is what the existing comment in Type_holder::aggregate_attributes already documents).

File: sql/sql_tvc.cc [hunk 2]
```diff
@@ -301,6 +345,22 @@ bool table_value_constr::prepare(THD *thd, SELECT_LEX *sl,
...
+      instances so that the types reported for the TVC reflect the values
+      seen on the current execution.
+    */
+    List_iterator<Item> it(sl->item_list);
+    for (uint pos= 0; pos < cnt; pos++)
+    {
+      Item_type_holder *ith= static_cast<Item_type_holder*>(it++);
+      ith->set_handler(holders[pos].type_handler());
+      ith->Type_std_attributes::set(holders[pos]);
+      ith->set_maybe_null(holders[pos].get_maybe_null());
```
[the remark this answers is not in the document]
Comment (DaveGosselin-MariaDB) [reply sql/sql_tvc.cc]: Are you sure?  Type_std_attributes is Type_numeric_attributes (max_length, decimals, unsigned_flag) plus collation. It does not contain a maybe_null field because that lives on Item. Type_std_attributes::set(*other) does *this = *other, which only copies the Type_std_attributes child object.  Removing the second call would lose nullability propagation between executes.

File: sql/sql_tvc.cc [hunk 3]
```diff
@@ -269,17 +273,57 @@ bool table_value_constr::prepare(THD *thd, SELECT_LEX *sl,
...
+    bool args_err= false;
+    for (uint pos= 0; pos < cnt && !args_err; pos++)
+    {
+      if (holders[pos].alloc_arguments(thd, lists_of_values.elements))
+        args_err= true;
+    }
+    if (arena)
+      thd->restore_active_arena(arena, &backup);
+    if (args_err)
+      DBUG_RETURN(true);
```
[the remark this answers is not in the document]
Comment (DaveGosselin-MariaDB) [reply sql/sql_tvc.cc]: Agreed, this is a bug, will update the patch with a fix.

File: sql/sql_tvc.cc [hunk 4]
```diff
@@ -269,17 +273,68 @@ bool table_value_constr::prepare(THD *thd, SELECT_LEX *sl,
   if (fix_fields_for_tvc(thd, li))
     DBUG_RETURN(true);
 
+  /*
+    Should be nullptr only during first PREPARE at which time it's
```
Comment (mariadb-YuchenPei) [inline sql/sql_tvc.cc]: > first PREPARE

I guess should be "PREPARE or first EXECUTE", or are there scenarios of multiple PREPAREs where it is not nullptr during the second and subsequent PREPAREs?

File: mysql-test/main/table_value_constr.result [hunk 5]
```diff
@@ -3279,3 +3279,41 @@ drop table t1,t2;
 #
 # End of 10.4 tests
 #
+#
+# MDEV-35747 Prepared statement with WITH ... AS (VALUES ... ) returns
+#            wrong result
+#
+# CTE whose body is a TVC with a parameter marker and an explicit column.
```
Comment (mariadb-YuchenPei) [inline mysql-test/main/table_value_constr.result:3302]: > CTE

It looks like the bug could be reprod at base (and is fixed at your patch) without CTE:

```
prepare stmt from 'values (?)';
set @val= 1;
execute stmt using @val;
deallocate prepare stmt;
```

Can you add a simplified testcase like this?

File: sql/sql_tvc.cc [hunk 6]
```diff
@@ -269,17 +273,68 @@ bool table_value_constr::prepare(THD *thd, SELECT_LEX *sl,
...
+      thd->restore_active_arena(arena, &backup);
+    if (args_err)
+      DBUG_RETURN(true);
+
+    holders= type_holders= new_holders;
+  }
+
+  /*
+    Collect the column types from the list of values.  This must run
+    on every PREPARE, not just the first one, so that values whose
```
Comment (mariadb-YuchenPei) [inline sql/sql_tvc.cc]: > This must run on every PREPARE,

Do you mean run on every EXECUTE?

**Comment by mariadb-YuchenPei** [comment]
claude review

---

# Code Review: MDEV-35747 — Wrong result from prepared TVC with parameter markers

**Author:** Dave Gosselin · **Files:** `sql/sql_tvc.cc` (+80/−9), `sql/sql_class.h` (+7 comment), tests (+65)

## Overview

A table value constructor used in a prepared statement (e.g. `WITH t(id) AS (VALUES (?)) SELECT ...`) returned stale results across `EXECUTE`s. Root cause: in `table_value_constr::prepare()`, the **type collection** (`join_type_handlers_for_tvc` + `get_type_attributes_for_tvc`) and the **`Item_type_holder` list build** were both wrapped in an `if (!holders)` guard that runs only once per statement. A parameter marker has no real type at PREPARE, so the TVC froze whatever placeholder type it had and never updated when a later `EXECUTE` bound a value of a different type.

The fix correctly splits the work:
- **Once per statement** (`if (!holders)`): allocate the `Type_holder[]` array *and* each holder's `args[]` buffer on the statement arena.
- **Every EXECUTE**: recompute type handlers/attributes, and either build the `Item_type_holder` list (first time) or mutate the existing instances in place (subsequent times). `get_type_attributes_for_tvc` now calls `remove_arguments()` (reset `arg_count`) instead of reallocating `args[]` each call.

## Correctness — verified

- ✅ **`args[]` reuse is safe.** `alloc_arguments(thd, lists_of_values.elements)` sizes each buffer to the (SQL-fixed, constant) row count; per-execute `remove_arguments()` + one `add_argument()` per row refills it exactly. `add_argument` has no bounds check, but the fill count equals capacity every time, so no overflow. The `count<=2` inline `tmp_arg` path also persists correctly since the `Type_holder` array lives on the stmt arena.
- ✅ **Types are genuinely recomputed per execute.** `join_type_handlers_for_tvc` does `set_handler()` on the first row each call (fresh, not accumulated), and `Item_hybrid_func_fix_attributes` recomputes `max_length`/`decimals`/collation into the holder. The test (`1` → `'abc'`, `10/20` → `'aa'/'bb'`) confirms the observable fix.
- ✅ **In-place mutation mirrors the constructor** for the type-varying fields: `set_handler` ↔ `Type_handler_hybrid_field_type(handler)`, `Type_std_attributes::set` ↔ ctor's `Type_std_attributes::set`, `set_maybe_null` ↔ ctor. `name` and the copied flags are structural and legitimately not re-set.
- ✅ **Contract upheld.** `get_type_attributes_for_tvc` has exactly one caller (`prepare`), and `args[]` is always allocated before it runs, in both first-prepare and re-execute paths.
- ✅ **Good defensive change:** building into a local `new_holders` and only publishing to `holders`/`type_holders` after all `alloc_arguments` succeed avoids leaving a half-initialized array behind on OOM — an improvement over the previous immediate assignment.

## Issues / risks

- **`enum_set_typelib` not refreshed in the mutate branch (minor completeness gap).** The constructor sets `enum_set_typelib(attr->get_typelib())`, but the `else` branch updates only handler/attrs/maybe_null. For parameter markers the aggregated typelib is `NULL`/constant, so there's no impact on this fix; but if a TVC column's aggregated type could ever resolve to ENUM/SET and change across executes, the cached typelib would go stale. Consider mirroring it for parity (needs a setter on `Item_type_holder`, whose `enum_set_typelib` is currently `protected`).
- **Sticky `maybe_null` (documented).** The new `sql_class.h` comment honestly notes that `aggregate_attributes` OR-accumulates `m_maybe_null` and never resets it, so a column stays nullable for the statement's lifetime once any execute sees NULL. This is **safe** (only over-widens nullability — no wrong result) and, for parameter-based TVCs, arguably correct anyway (an unbound `?` is nullable). It is, however, inconsistent with the handler, which *is* reset each execute. If you want tightness, resetting `m_maybe_null= false` at the top of `aggregate_attributes` (or right after `remove_arguments`) would make nullability reflect only the current execute. Documenting-instead-of-fixing is a reasonable call here.
- **Unchecked `static_cast<Item_type_holder*>` in the mutate branch.** Safe today because that list is built exclusively of `Item_type_holder`s, but a `DBUG_ASSERT((*it)->type() == Item::TYPE_HOLDER)` would harden against future changes.

## Style

- Clear, well-scoped comments explaining *why* each block is once-vs-per-execute — genuinely helpful.
- **Minor:** the "already built" discriminator `sl->item_list.elements != cnt` reads slightly obliquely; `sl->item_list.is_empty()` would express "not yet built" more directly (equivalent, since the build always produces exactly `cnt`).
- Two separate `activate_stmt_arena_if_needed`/`restore_active_arena` pairs in the same prepare — fine and necessary given the split, just slightly repetitive.

## Test coverage

- ✅ Solid: single param, multi-row/multi-param with `ORDER BY`, type change across executes (int→string), and repeated identical executes (regression against the frozen-type bug).
- **Suggestions:** add (a) a **NULL-then-non-NULL** bind to lock in the documented sticky-nullable behavior; (b) a **multi-column** VALUES with parameter markers (`VALUES (?, ?)` as columns, not rows) to exercise per-column holders; both are cheap and cover paths the current tests don't.

## Security

- No concerns. No new external input handling; buffer reuse is bounded by the fixed row count; allocations are arena-scoped.

## Verdict

**Approve.** Correct, well-reasoned fix that properly separates per-statement from per-execute work and eliminates a real wrong-results bug, with a nice defensive improvement around partial initialization. Only minor follow-ups worth considering: refresh `enum_set_typelib` in the mutate path for parity, optionally reset `m_maybe_null` for tighter nullability, add the `static_cast` assert, and extend tests with NULL-binding and multi-column cases.

File: sql/sql_tvc.cc [hunk 7]
```diff
@@ -269,17 +273,68 @@ bool table_value_constr::prepare(THD *thd, SELECT_LEX *sl,
...
+      get_type_attributes_for_tvc(thd, li, holders,
+                                  lists_of_values.elements, cnt))
+     DBUG_RETURN(true);
+
+  /*
+    If the counts don't match, then allocate some more
+    Item_type_holder instances.  These will, on subsequent EXECUTEs,
+    be mutated in place (see the 'else' case following).
+  */
+  if (sl->item_list.elements != cnt)
```
Comment (mariadb-YuchenPei) [inline sql/sql_tvc.cc]: With these two assertions the tests main.opt_tvc and table_value_constr still pass:

```modified   sql/sql_tvc.cc
@@ -335,6 +335,9 @@ bool table_value_constr::prepare(THD *thd, SELECT_LEX *sl,
   */
   if (sl->item_list.elements != cnt)
   {
+    DBUG_ASSERT(sl->item_list.elements == 0);
+    DBUG_ASSERT(thd->stmt_arena->is_stmt_prepare_or_first_stmt_execute() ||
+                thd->stmt_arena->is_conventional());
     List_iterator_fast<Item> it(*first_elem);
     Item *item;
     Query_arena *arena, backup;

```

Do you think we can simplify the if condition to the one in the first assert (or "sl->item_list.is_empty()" as claude suggests), and add the second assert which also appears in the "if (!holders)" above?

File: sql/sql_tvc.cc [hunk 8]
```diff
@@ -269,17 +273,68 @@ bool table_value_constr::prepare(THD *thd, SELECT_LEX *sl,
...
                 thd->stmt_arena->is_conventional());
-    holders= type_holders=
+    /*
+      Build the Type_holder array in a local first.  Only save it to
+      the persistent type_holders member once every holder has its
+      args[] buffer in place.  If alloc_arguments() fails partway
+      through, then we must not leave type_holders pointing at a
+      partially initialized array.  A later call would then take the
+      "already initialized" path, skip this block, and dereference a
+      NULL args[] inside add_argument().
```
Comment (mariadb-YuchenPei) [inline sql/sql_tvc.cc]: I don't understand the last bit of this comment, and "dereference a NULL args[]" sound like something to avoid... Can you make it a bit more readable?

File: sql/sql_tvc.cc [hunk 9]
```diff
@@ -301,6 +356,22 @@ bool table_value_constr::prepare(THD *thd, SELECT_LEX *sl,
     if (unlikely(thd->is_fatal_error))
       DBUG_RETURN(true); // out of memory
   }
+  else
+  {
+    /*
+      Refresh the cached type info on the existing Item_type_holder
+      instances so that the types reported for the TVC reflect the values
+      seen on the current execution.
```
Comment (mariadb-YuchenPei) [inline sql/sql_tvc.cc:373]: As claude remarked in its review, the three set statements below nicely mirror the Item_type_holder constructor. Perhaps point that out here in the comment? Otherwise "ith->Type_std_attributes::set" looks quite complex

File: sql/sql_tvc.cc [hunk 10]
```diff
@@ -269,17 +273,68 @@ bool table_value_constr::prepare(THD *thd, SELECT_LEX *sl,
...
     DBUG_RETURN(true);
 
+  /*
+    Should be nullptr only during first PREPARE at which time it's
+    then allocated and will be set for the rest of the statement's
+    lifetime.
+  */
   if (!holders)
   {
     DBUG_ASSERT(thd->stmt_arena->is_stmt_prepare_or_first_stmt_execute() ||
```
Comment (mariadb-YuchenPei) [inline sql/sql_tvc.cc:291]: I tried to narrow the "is_stmt_prepare_or_first_stmt_execute()" to "is_stmt_prepare()" and I get assertion failures in SP testcases as well as PS cases such as "SELECT * FROM t1 WHERE c IN ('10','20')" as the prepared statement. So for some reason the it could be the first EXECUTE too, not just PREPARE. Can you update the commit message to reflect that? Currently it refers to this branch as PREPARE only.

**Review by mariadb-YuchenPei — CHANGES_REQUESTED** [review]

File: sql/sql_tvc.cc (hunk 4 above)
Comment (DaveGosselin-MariaDB) [reply sql/sql_tvc.cc]: Thanks, right, I reworded the comment; I can't see how it would be otherwise.

File: mysql-test/main/table_value_constr.result (hunk 5 above)
Comment (DaveGosselin-MariaDB) [reply mysql-test/main/table_value_constr.result:3302]: Great catch, I added a test case similar to this where it attempts to execute multiple times and with multiple different types before deallocating.

File: sql/sql_tvc.cc (hunk 7 above)
Comment (DaveGosselin-MariaDB) [reply sql/sql_tvc.cc]: Yes, simplified.

File: sql/sql_tvc.cc (hunk 9 above)
Comment (DaveGosselin-MariaDB) [reply sql/sql_tvc.cc:373]: I decided to factor the common bits to a new method on the class and call that from this code as well as the class constructor.

File: sql/sql_tvc.cc (hunk 10 above)
Comment (DaveGosselin-MariaDB) [reply sql/sql_tvc.cc:291]: Added another paragraph to the commit message to clarify.

File: sql/sql_tvc.cc (hunk 8 above)
Comment (DaveGosselin-MariaDB) [reply sql/sql_tvc.cc]: I reworded the comment for clarity.

File: sql/sql_tvc.cc (hunk 6 above)
Comment (DaveGosselin-MariaDB) [reply sql/sql_tvc.cc]: Yes, thanks.

File: sql/sql_tvc.cc (hunk 4 above)
Comment (mariadb-YuchenPei) [reply sql/sql_tvc.cc]: It still says "first PREPARE". Are there cases of multiple PREPAREs where it is not nullptr during the second and subsequent PREPAREs?

File: sql/sql_tvc.cc (hunk 10 above)
Comment (mariadb-YuchenPei) [reply sql/sql_tvc.cc:291]: Thanks. The new paragraph says "For a prepared statement it happens at PREPARE". But there are (edge?) cases where it can happen at the first EXECUTE, such as when a TVC is constructed due to the "IN predicates into subqueries" optimisation e.g. this block in table_value_constr.test:

```
--echo #
--echo # MDEV-23182: Server crashes in
--echo # Item::fix_fields_if_needed / table_value_constr::prepare upon 2nd execution of PS
--echo #
SET @save_in_predicate_conversion_threshold=@@in_predicate_conversion_threshold;
SET in_predicate_conversion_threshold=2;

CREATE TABLE t1 (c VARCHAR(10)) DEFAULT CHARSET=utf8;
PREPARE stmt FROM "SELECT * FROM t1 WHERE c IN ('10','20')";
EXECUTE stmt;
--echo # Without the patch second execution of the prepared statement 'stmt'
--echo # results in crash.
EXECUTE stmt;
DEALLOCATE PREPARE stmt;

DROP TABLE t1;
```

File: sql/sql_tvc.cc [hunk 11]
```diff
@@ -269,17 +273,71 @@ bool table_value_constr::prepare(THD *thd, SELECT_LEX *sl,
...
+  if (join_type_handlers_for_tvc(thd, li, holders, cnt) ||
+      get_type_attributes_for_tvc(thd, li, holders,
+                                  lists_of_values.elements, cnt))
+     DBUG_RETURN(true);
+
+  /*
+    If the counts don't match, then allocate some more
+    Item_type_holder instances.  These will, on subsequent EXECUTEs,
+    be mutated in place (see the 'else' case following).
+  */
```
Comment (mariadb-YuchenPei) [inline sql/sql_tvc.cc:344]: The comment needs update too now that the condition has been simplified. Suggest something like "Populate the item_list with the Item_type_holders just constructed"

File: sql/sql_tvc.cc [hunk 12]
```diff
@@ -269,17 +273,71 @@ bool table_value_constr::prepare(THD *thd, SELECT_LEX *sl,
...
+  if (sl->item_list.is_empty())
+  {
+    DBUG_ASSERT(thd->stmt_arena->is_stmt_prepare_or_first_stmt_execute() ||
+                thd->stmt_arena->is_conventional());
```
Comment (mariadb-YuchenPei) [inline sql/sql_tvc.cc]: Given this assert also appears in the "if (!holder)" branch above, I wonder if this "if" branch is entered if and only if that "if" branch is. If that is the case, perhaps it would be a good idea to make this connection clear, with a comment or a boolean set to "!holder" above

**Comment by mariadb-YuchenPei** [comment]
claude review comment:

---

# Code Review: MDEV-35747 — Wrong result from prepared TVC with parameter markers (revised)

**Author:** Dave Gosselin · **Files:** `sql/sql_tvc.cc` (+85/−9), `sql/item.h` (+11/−3), `sql/sql_class.h` (+7 comment), tests (+92)

## Overview

Same core fix as the prior revision: `table_value_constr::prepare()` computed TVC column types only once (under an over-broad `if (!holders)` guard), freezing a parameter marker's placeholder type at build time so later `EXECUTE`s with a different type returned wrong results/metadata. The fix splits **once-per-statement** work (allocate the `Type_holder[]` array + each holder's `args[]` buffer on the statement arena) from **per-execute** work (recompute handlers/attributes, then build-or-refresh the `Item_type_holder` list).

## What changed since the previous revision — and it's better

- ✅ **`refresh()` helper extracted** on `Item_type_holder` (`item.h`), called by *both* the constructor and the per-execute mutate path. This eliminates the constructor-vs-mutation drift risk I raised before: handler, std attributes, and `maybe_null` now provably stay in sync by construction. Verified it compiles — the default `Type_handler_hybrid_field_type()` ctor exists (`sql_type.h:7569`), `set_handler()` replaces the old init-list handler, and `Type_std_attributes::set(const Type_std_attributes*)` (pointer overload at `sql_type.h:3225`) binds `attr`/`&holders[pos]`.
- ✅ **Discriminator is now `sl->item_list.is_empty()`** instead of `elements != cnt` — clearer "not yet built" intent, as suggested.
- ✅ **`DBUG_ASSERT(elem->type() == Item::TYPE_HOLDER)`** added before the `static_cast` in the mutate branch — the hardening I recommended.
- ✅ **New bare-`VALUES (?)` test** (no CTE) added, exercising a TVC as the top-level prepared statement — good extra coverage.
- ✅ **Clearer commit message** explaining that first allocation coincides with PREPARE for a PS but with first execution for a stored procedure, which is why the assertion permits both.

## Correctness — verified (carried over + re-checked)

- ✅ **`args[]` reuse is safe:** one-time `alloc_arguments(thd, lists_of_values.elements)` sizes each buffer to the SQL-fixed row count; per-execute `remove_arguments()` + exactly-one `add_argument()` per row refills it to capacity (no bounds check in `add_argument`, but the fill count equals capacity every time).
- ✅ **Types genuinely recomputed per execute:** `join_type_handlers_for_tvc` re-`set_handler()`s on the first row (fresh, not accumulated), and `aggregate_attributes`→`Item_hybrid_func_fix_attributes` recomputes std attributes each call.
- ✅ **`holders`/`item_list` stay coupled:** both persist on the stmt arena and are built together on the first call; the only `item_list.empty()` is inside the guarded build block, so nothing re-empties it per execute (no per-execute allocation growth).
- ✅ **Defensive local `new_holders`** published to `type_holders` only after all `args[]` allocs succeed — avoids a half-initialized array being seen as "already built."
- ✅ **`is_empty()` build branch keeps the arena assertion**, consistent with first-time (PREPARE or first SP execution) build.

## Remaining minor points (mostly unchanged from prior review)

- **`enum_set_typelib` still not refreshed.** The constructor sets `enum_set_typelib(attr->get_typelib())`, but `refresh()` — which now reads as "refresh all the type info" — omits it, so the mutate path never updates it. For parameter markers the aggregated typelib is `NULL`/constant, so no bug in the target scenario; but the omission is now easier to trip over given `refresh()`'s name. Recommend either folding typelib into `refresh()` (and having the ctor set it via `refresh` too) or adding a one-line comment on why it's deliberately excluded.
- **Sticky `maybe_null` (documented).** The `sql_class.h` comment honestly records that `m_maybe_null` OR-accumulates and never resets, so a column stays nullable for the statement's lifetime once any execute binds NULL. Safe direction (over-nullable only, no wrong result) and, for `?`-based TVCs, effectively expected. Documenting rather than resetting is a reasonable call; a `m_maybe_null= false` reset at the top of `aggregate_attributes` would make it precise if desired.

## Test coverage

- ✅ Good: bare `VALUES (?)`, CTE-wrapped TVC, and multi-row/multi-param with `ORDER BY`; each repeats the same bind then switches type (int→string) to guard the frozen-type regression.
- **Still worth adding:** (a) a **NULL-then-non-NULL** bind to lock in the documented sticky-nullable behavior; (b) a **multi-column** `VALUES (?, ?)`-as-columns case (current multi-param test is multi-row, single-column) to exercise per-column holders independently.

## Security / performance

- No concerns. Buffer reuse is bounded by the fixed row count; allocations are arena-scoped and now happen once instead of per-execute (a small perf improvement over pre-fix behavior); no new external-input handling.

## Verdict

**Approve.** This revision keeps the correct once-vs-per-execute split and improves on the previous version: the shared `refresh()` removes the parity risk, `is_empty()` and the type assertion address earlier nits, and the bare-`VALUES` test broadens coverage. Only optional follow-ups remain — decide on `enum_set_typelib` in `refresh()` (fold in or comment), optionally reset `m_maybe_null` for tighter nullability, and add the NULL-bind and multi-column tests.

**Review by mariadb-YuchenPei — CHANGES_REQUESTED** [review]
~LGTM after addressing review comments.~ Sorry actually I'd like to take another look after this round.

Can you please also address claude's review comments - the "**`enum_set_typelib` still not refreshed.**" one and the "**Still worth adding:**" one?

**Comment by mariadb-YuchenPei** [comment]
claude:

---

Reviewed `db0d0d4e` (MDEV-35747, TVC type re-computation on every EXECUTE). The core split of "once per statement" vs "once per execution" work looks sound, and I verified the supporting machinery: `unit->reinit_exec_mechanism()` does reset `prepared` so `table_value_constr::prepare()` really is re-entered on each EXECUTE; `join_type_handlers_for_tvc()` fully re-seeds the handler from the first row; `active_stmt_arena_to_use()` and `activate_stmt_arena_if_needed()` agree on which arena to use, so the `Type_holder[]` array and the `args[]` buffers land on the same mem_root; `Item_args::remove_arguments()` only zeroes `arg_count`, and the `args[]` capacity (`lists_of_values.elements`) exactly matches the number of `add_argument()` calls per holder; `copy_flags()` masks, so it does not clobber the MAYBE_NULL bit set by the new `refresh()`.

Two findings:

- `sql/sql_tvc.cc:378` — NULL dereference on re-execution when the first prepare fails after `type_holders` was assigned. `type_holders` now doubles as the flag for "`sl->item_list` is populated", but it is set at line 326 while the item list is built at 345-367. If `join_type_handlers_for_tvc()`/`get_type_attributes_for_tvc()` errors (e.g. `CREATE PROCEDURE p() VALUES (1,POINT(1,1)),(1,1);` — first `CALL` raises ER_ILLEGAL_PARAMETER_DATA_TYPES2_FOR_OPERATION), the item list stays empty; the second `CALL` takes the `else` branch, `it++` returns NULL and `elem->type()` crashes. Same shape via a collation-aggregation error in an SP, and via OOM where `push_back(NULL)` stores a NULL before line 365 returns. Pre-patch, `holders` was never read outside the `if (!holders)` block, so this was not reachable.

- `sql/sql_class.h:8264` — reused `Type_holder` never resets `m_maybe_null` (nor `m_typelib`). Now that `aggregate_attributes()` runs on every execution against the same holders, the OR-accumulation is sticky: `CREATE PROCEDURE p(a INT) VALUES (a); CALL p(NULL); CALL p(1);` reports the column as NULLable on the second call. The patch adds a comment describing this rather than clearing the flag; resetting it alongside `remove_arguments()` in `get_type_attributes_for_tvc()` would make the recomputation idempotent.

Minor, not reported as findings: `get_type_attributes_for_tvc()`'s `count_of_lists` parameter is now unused (its doc comment still describes it), and the function has silently acquired a precondition that the caller pre-allocated each holder's `args[]` — there is only one caller today, but `add_argument()` would write through a NULL `args` if that ever changed.

**Review by mariadb-YuchenPei — COMMENTED** [review]
I don't have comments from myself.

Claude is not saying anything new, except

1. one of the points it raised is the same as copilot and worth addressing
2. unused argument count_of_lists could be removed too

Regarding the Point 1, here's a testcase and it crashes on the second CALL

```sql
SET @a= _latin1 'a';
SET @b= _latin2 'b';
CREATE PROCEDURE p() VALUES (@a),(@b);
--error ER_CANT_AGGREGATE_2COLLATIONS
CALL p;               # fails in get_type_attributes_for_tvc(), item_list left empty
SET @b= _latin1 'b';  # now both latin1
CALL p;               # first_call==false, it++ returns NULL -> crash at line 378/380
DROP PROCEDURE p;```

**Comment by mariadb-YuchenPei** [comment]
claude:

---

This version addresses the points from the earlier review:
- **Crash after a failed first call:** it now decides whether to build the SELECT list by checking `sl->item_list.is_empty()`, so the empty-list crash is gone. The latin1/latin2 SP test covers it.
- **Nullability:** it is reset in `get_type_attributes_for_tvc()`, and the test now expects N, Y, N.
- **Cleanup:** the unused `count_of_lists` is removed.
- **`sql/item.cc` change:** `in_tvc` is true only for the duration of `table_value_constr::prepare()`, and `current_select` is the TVC's select at that point. So skipping the stmt-arena branch there just turns the conversion into an ordinary `change_item_tree()` that gets rolled back each execution. That's correct, and it stops the stmt mem_root growing on every EXECUTE.

Remaining findings:

1. **Out-of-memory can still leave the list half-built** (`sql/sql_tvc.cc`, the `is_empty()` block). If an `Item_type_holder` allocation fails, `push_back(NULL)` puts a NULL into the list. If `push_back` itself fails, the list ends up shorter than `cnt`. Either way the list is no longer empty, so the next call takes the `else` branch and `it++` returns NULL, then `elem->type()` crashes. It only happens after OOM, so it's low severity. The fix is to build into a local `List<Item>` and assign it to `sl->item_list` only on success, or to empty `sl->item_list` before the OOM return.

2. **The IN-predicate→TVC path has no test.** The commit message names it as one of the two cases where the first `prepare()` happens at the first EXECUTE, but every new test is either a PS with a parser TVC or an SP. It needs a PS like `x IN (?,?,…)` with `in_predicate_conversion_threshold` lowered, executed with non-NULL values and then with a NULL, and with values of different types. This is also where column nullability now changes between executions. Anything decided once at the first execution from the subquery column's nullability, such as the semijoin/IN-to-EXISTS setup, is worth checking against a later NULL.

3. **`Type_holder::m_typelib` is still never reset** (`sql/sql_class.h`). `refresh()` copies it into `enum_set_typelib` on every execution, so an ENUM typelib from one execution would carry over into a later execution of another type. I can't find a way to reach this today: a TVC value's type only changes between executions through `?` or `@var`, and neither can be ENUM/SET. For symmetry it could be cleared next to `set_type_maybe_null(false)`.

---

(claude later withdrew finding 3 when prompted for a testcase because "the stale typelib has no visible effect")

File: sql/item.cc [hunk 13]
```diff
@@ -2733,11 +2733,20 @@ bool Type_std_attributes::agg_item_set_converter(const DTCollation &coll,
     if (conv->fix_fields_if_needed(thd, arg))
       return TRUE;
 
+    /*
+      A table value constructor aggregates the types of its values again on
+      every execution, and its query block never clears
+      first_cond_optimization because only JOIN optimization does that.
+      Without the in_tvc test every execution would add a conversion to the
+      statement memory root, which must not grow once the statement has
+      run.
```
Comment (mariadb-YuchenPei) [inline sql/item.cc]: This comment is a bit odd, because conv has already been created on
mem_root regardless of which branch is entered. Suggest change it to
something like:

```
    For A table value constructor every execution would allocate a new
    Item_direct_ref_to_item on the statement mem_root, because
    Type_holder::args[] is refilled from the original values each time and
    first_cond_optimization is never cleared for a TVC.
```

File: sql/sql_tvc.cc [hunk 14]
```diff
@@ -269,23 +278,74 @@ bool table_value_constr::prepare(THD *thd, SELECT_LEX *sl,
...
+      get_type_attributes_for_tvc(thd, li, holders, cnt))
+     DBUG_RETURN(true);
+
+  /*
+    Populate the SELECT list with the Item_type_holder instances that
+    describe the columns of the TVC.  On every later execution these same
+    instances are refreshed in place (see the 'else' case following).
+    The list says whether they are there, not the holder array.  The type
+    collection above runs between the two and can report an error, which
+    leaves the holder array allocated and the list still empty.
```
Comment (mariadb-YuchenPei) [inline sql/sql_tvc.cc]: The comment is a bit obfuscated. Suggest rewording "The list says..." to:

```
    The list says whether they are there, not the holders array. The
    type collection above runs after populating the holders and before
    populating the list, so if it reports an error, it will leave the
    holders array allocated but the list still empty.
```

**Review by mariadb-YuchenPei — COMMENTED** [review]
LGTM (pass to @spetrunia) once the two comments on code comments are addressed.

Claude comments are all nits. I think only Point 2 is good to have
